src/sparkle/application/__init__.py

```diff
@@ -1,95 +1,163 @@
-
-    Attributes:
-        spark_session (SparkSession): The Spark session used for
-          executing Spark jobs.
-        config (Config): Configuration object containing
-          application-specific settings.
-        sources (Dict[str, Source]): Dictionary of data sources used
-          in the application, keyed by source name.
-        writers (List[Writer]): List of Writer objects used for
-          writing the processed data to various destinations.
-
+    This class provides a foundation for Spark-based data processing applications.
+    It handles setting up the Spark session, configuring readers and writers,
+    and defining the main process logic through abstract methods that must be implemented
+    by subclasses.
     """
 
     def __init__(
         self,
-        spark_session: SparkSession,
         config: Config,
+        readers: dict[str, type[Reader]],
         writers: list[Writer],
+        spark_extensions: list[str] | None = None,
+        spark_packages: list[str] | None = None,
+        extra_spark_config: dict[str, str] | None = None,
     ):
-        """Sparkle's application initializer.
+        """Initializes the Sparkle application with the given configuration.
 
         Args:
-            spark_session (SparkSession): The Spark session used for
-              executing Spark jobs.
-            config (Config): Configuration object containing
-              application-specific settings.
-            sources (Dict[str, Source]): Dictionary of data sources
-              used in the application, keyed by source name.
-            writers (List[Writer]): List of Writer objects used for
-              writing the processed data to various destinations.
+            config (Config): The configuration object containing application-specific settings.
+            readers (dict[str, type[Reader]]): A dictionary of readers for input data, keyed by source name.
+            writers (list[Writer]): A list of Writer objects used to output processed data.
+            spark_extensions (list[str], optional): A list of Spark session extensions to apply.
+            spark_packages (list[str], optional): A list of Spark packages to include in the session.
+            extra_spark_config (dict[str, str], optional): Additional Spark configurations to
+              merge with the default settings.
         """
-        self.spark_session = spark_session
         self.config = config
         self.writers = writers
+        self.readers = readers
+        self.execution_env = config.execution_environment
+        self.spark_config = config.get_spark_config(
+            self.execution_env, extra_spark_config
+        )
+        self.spark_extensions = config.get_spark_extensions(spark_extensions)
+        self.spark_packages = config.get_spark_packages(spark_packages)
+
+        self.spark_session = self.get_spark_session(self.execution_env)
+
+    def get_spark_session(self, env: ExecutionEnvironment) -> SparkSession:
```

Comment:
```suggestion
    def get_spark_session(self, env: ExecutionEnvironment | None) -> SparkSession:
```

Reply:
I should apply these if it's ok for Sparkle to accept a Spark session; let's see what @shahinism suggests.